### DISCUSSION TITLE
How to make `KP_` keys act the same way as the default ones?

### DISCUSSION THREAD

**Discussion by marchersimon** [i3/i3#4675] [Q&A]
Using the keypad produces `KP_` key events. For example, typing `1` on the keypad produces `KP_1`. i3 differentiates between those two. In the config file, you can just specify `KP_1` as key and it works, but this behavior also applies to all terminal emulators running in i3 (at least for `urxvt` and `xterm`). 
While in the prompt, `KP_1` doesn't do anything, but when running any interactive tool in the terminal, it works. It even works when running `sleep 10` and typing while it runs. For any other app (`dmenu`, `firefox`, `emacs`, …) it works without a problem.

Is there any way to get this working at least in the terminal? Thanks.

**Comment by Airblader** [comment]
Can you describe things with a specific example rather than in generic terms? I'm not at all following what you're saying, sorry.

What keysym a physical key produces has nothing with i3, however, or whether an application was launched through i3.

**Comment by marchersimon** [comment]
When I press any key on the numpad like `1`, it produces `KP_1`. And i3 differentiates between `KP_1` and `1`. That's all fine, because I can just use `KP_1` in my config file. 
But my problem is, that for some reason, it also doesn't work in `urxvt` and `xterm`. When I press `KP_1`, it doesn't do anything. When using `urxvt` and `xterm` in KDE however, it works as expected and that's why I think it has to do with `i3`.

In both `urxvt` and `xterm`, it only doesn't work in the prompt. When opening any interactive tool I can write `KP_1` and I get a `1`.

**Reply by Airblader** [reply]
Did you make sure NumLock is enabled?

**Reply by marchersimon** [reply]
Yes, otherwise it also wouldn't work in other applications

**Comment by netyaroze** [comment]
How are you setting the KP_* keys in your configuration? i3 isn't really smart enough to figure out that Numlock is on by itself. 
I would need to have a look at that try and replicate it.

I dont use a numpad anymore, but when i did i was defining (mod2 being numlock for people unfamiliar):
`$mod+mod2+KP_1`

Also, if you haven't seen already, there may be some more clarity in this thread for you:
https://github.com/i3/i3/issues/2558#issuecomment-405577744

**Reply by Airblader** [reply]
> i3 isn't really smart enough to figure out that Numlock is on by itself

This comment feels a bit unfair. Keyboard management is really difficult, and XKB is a mess. If you want to have some fun, take a look at https://github.com/i3/i3/commit/bf3cd41b5ddf1e757515ab5fbf811be56e5f69cc. It's not about being smart, but about being correct.

**Reply by netyaroze** [reply]
Apologies :) I meant no disrespect in that regard. Obviously should have provided a more technical response to cover that (albeit 5mins in my lunch break doesn’t let me do that sometimes). Definitely agree.
